src/compass/s1_geocode_slc.py

```diff
@@ -83,55 +75,69 @@ def run(cfg: GeoRunConfig):
-            f'{burst_output_path}/{id_pol}.slc',
-            geo_grid.width, geo_grid.length,
-            rdr_burst_raster.num_bands, gdal.GDT_CFloat32,
-            cfg.geocoding_params.output_format)
-
         # Extract burst boundaries
         b_bounds = np.s_[burst.first_valid_line:burst.last_valid_line,
-                   burst.first_valid_sample:burst.last_valid_sample]
+                         burst.first_valid_sample:burst.last_valid_sample]
 
         # Create sliced radar grid representing valid region of the burst
         sliced_radar_grid = burst.as_isce3_radargrid()[b_bounds]
 
-        # Geocode
-        isce3.geocode.geocode_slc(geo_burst_raster, rdr_burst_raster,
-                                  dem_raster,
-                                  radar_grid, sliced_radar_grid,
-                                  geo_grid, orbit,
-                                  native_doppler,
-                                  image_grid_doppler, ellipsoid, threshold,
-                                  iters, blocksize, flatten,
-                                  azimuth_carrier=az_carrier_poly2d)
-
-        # Set geo transformation
-        geotransform = [geo_grid.start_x, geo_grid.spacing_x, 0,
-                        geo_grid.start_y, 0, geo_grid.spacing_y]
-        geo_burst_raster.set_geotransform(geotransform)
-        geo_burst_raster.set_epsg(epsg)
-        del geo_burst_raster
-        del dem_raster # modified in geocodeSlc
-
-        # Save burst metadata
-        metadata = GeoCslcMetadata.from_georunconfig(cfg, burst_id)
-        json_path = f'{burst_output_path}/{id_pol}.json'
-        with open(json_path, 'w') as f_json:
-            metadata.to_file(f_json, 'json')
+        output_hdf5 = out_paths.hdf5_path
+        with h5py.File(output_hdf5, 'w') as geo_burst_h5:
+            backscatter_group = geo_burst_h5.require_group('complex_backscatter')
+            init_geocoded_dataset(backscatter_group, pol, geo_grid,
+                                  'complex64', f'{pol} geocoded SLC image')
+
+            # access the HDF5 dataset for a given frequency and polarization
+            dataset_path = f'/complex_backscatter/{pol}'
+            gslc_dataset = geo_burst_h5[dataset_path]
+
+            # Construct the output raster directly from HDF5 dataset
+            geo_burst_raster = isce3.io.Raster(f"IH5:::ID={gslc_dataset.id.id}".encode("utf-8"), update=True)
+
+            # Geocode
+            isce3.geocode.geocode_slc(geo_burst_raster, rdr_burst_raster,
+                                      dem_raster,
+                                      radar_grid, sliced_radar_grid,
+                                      geo_grid, orbit,
+                                      native_doppler,
+                                      image_grid_doppler, ellipsoid, threshold,
+                                      iters, blocksize, flatten,
+                                      azimuth_carrier=az_carrier_poly2d)
```

Comment:
Does this work if we decide to write complex float 32 (i.e. two float16)? 
Right now with raster interface we are letting gdal write to the HDF5, which is fine for complex 64. But I guess that would not work for comlex 32 (float 16). right? 
I assume we could change the geocode_slc interface here and use the array interface and have the flexibility to write what we wish at python level. But I guess we would need the array interface to be pushed out to public.

Reply:
Correct, it does not work for complex float 32. To support complex float 32, I think both modes of ISCE3 `geocode_slc` would need to be templated to support complex float 32.

I agree that the full python interface would give the most flexibility and ease in adopting compression.